### ISSUE TITLE
Persistent compose window timeouts

### ISSUE THREAD

**Issue by jismay-cenic** [roundcube/roundcubemail#5719]
Some of my users are seeing timeouts in the compose window on a regular basis. I have been unable to replicate at will so far.

General server configurations:

CentOS 6.8
Apache: httpd-2.2.15-55.el6.centos.2.x86_64
PHP: php-5.3.3-48.el6_8.x86_64
Roundcube: 1.2.3
MySQL: mysql-server-5.1.73-7.el6.x86_64

I have two servers which are behind a load-balancer. Load balancer is set to be persistent for 60 minutes.
Mysql is used for roundcube sessions. There are databases on both servers which are replicating master-master. No mysql errors have been noted.
Roundcube sessions have been set to "0" for to never expire.

Roundcube error logs show:

[21-Mar-2017 14:10:59 -0700]: <cufv5euo> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1490130603016&_lang=en_US&_framed=1&_action=send)
[21-Mar-2017 14:21:24 -0700]: <t963pnct> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1490131227943&_lang=en_US&_framed=1&_action=send)
[21-Mar-2017 14:36:17 -0700]: <tuqoghiq> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1490132120902&_lang=en_US&_framed=1&_action=send)
[21-Mar-2017 14:42:01 -0700]: <ev76v2lp> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1490132464766&_lang=en_US&_framed=1&_action=send)
[21-Mar-2017 14:42:16 -0700]: <ev76v2lp> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1490132479512&_lang=en_US&_framed=1&_action=send)
[21-Mar-2017 15:23:20 -0700]: <p4rpgp4u> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1490134943705&_lang=en_US&_framed=1&_action=send)
[03-Apr-2017 07:43:44 -0700]: <h3btcsmv> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1491230622848&_lang=en_US&_framed=1&_action=send)
[03-Apr-2017 07:43:52 -0700]: <h3btcsmv> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1491230629929&_lang=en_US&_framed=1&_action=send)
[03-Apr-2017 07:43:57 -0700]: <h3btcsmv> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1491230635280&_lang=en_US&_framed=1&_action=send)
[03-Apr-2017 07:44:11 -0700]: <h3btcsmv> PHP Error: Invalid compose ID in /var/www/html/roundcubemail-1.2.3/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1491230649857&_lang=en_US&_framed=1&_action=send)

**Comment by alecpl** [comment]
I think that's something with mysql replication or load balancer. When you access the compose page, it first stores some data to the session db under generated compose ID, then does redirect to the compose page using this ID and it looks it can't find that data in session.

BTW, I don't think session_lifetime=0 is well tested and supported. I propose to set it to some bigger value like 60 or more if you need, but not 0.

**Comment by jismay-cenic** [comment]
Noted on the session_lifetime setting. I have tried with it set to high values and 0 without any noticeable effects either way.
I will take a look at my mysql settings and see if that might shed any light on things.

**Comment by Maxfs79** [comment]
Hi!
We have a similar issue:
[07-Apr-2017 11:21:29 +0300]: <ate204tc> PHP Error: Invalid compose ID in /usr/local/www/roundcube/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1491553278222&_lang=ru&_framed=1&_action=send)

There are many similar posts in which it is not clear how to solve the issue. Our server configuration:
FreeBSD 10.3
nginx 1.10
php-fpm 5.6.30
mysql 5.6.35
roundcube 1.2.4

Perhaps there are some patches that fix these issue or tuning php settings?

Similar issue:
https://github.com/roundcube/roundcubemail/issues/5390
https://github.com/roundcube/roundcubemail/issues/4649

**Comment by alecpl** [comment]
> I think that's something with mysql replication or load balancer. When you access the compose page, it first stores some data to the session db under generated compose ID, then does redirect to the compose page using this ID and it looks it can't find that data in session.

To be more correct... From the provided errors, it looks that the compose data is in session on that initial redirect, but cannot be fond on 'send' action (it may be draft save on send action). So, it does not look much as a replication issue, but still something's wrong with the session. Are those users able to use Roundcube without re-login after the issue occurs?

**Comment by alecpl** [comment]
I'd also like to know if you're using ip_check=true. I think this could cause session destroy when user IP address has been changed while composing a message.

**Comment by jismay-cenic** [comment]
>>I think that's something with mysql replication or load balancer. When you access the compose page, it first stores some data to the session db under generated compose ID, then does redirect to the compose page using this ID and it looks it can't find that data in session.

> To be more correct... From the provided errors, it looks that the compose data is in the session on that initial redirect, but cannot be found on 'send' action (it may be draft save on send action). So, it does not look much as a replication issue, but still something's wrong with the session. Are these users able to use Roundcube without re-login after the issue occurs?

Yes, they are indeed able to use Roundcube without re-login. They can successfully close the failing compose window and perform all normal actions including a fresh compose without having to re-login.

>I'd also like to know if you're using ip_check=true. I think this could cause session destroy when user IP address has been changed while composing a message.

ip_check is currently set to false on all my servers.

-----

Since reverting session management to PHP-controlled rather than database my users are no longer reporting compose errors, and I am no longer seeing the above referenced log errors in the logs.

**Comment by sm-iitk** [comment]
I am also getting this error for few users with RC version 1.2.3 and default centoOS7 updated version.

PHP Error: Invalid compose ID in /var/www/html/program/steps/mail/sendmail.inc on line 39 (POST /?_task=mail&_unlock=loading1492595935810&_lang=en_GB&_framed=1&_action=send)

My IP_check=true, user did not change his IP during compose, his session was not expired.
I used following command, and the error went away, but a better fix is required.

# bin/deluser.sh <username>

**Comment by alecpl** [comment]
@sm-iitk so, this might depend on user preferences, next time this happens copy the user record before delete.

**Comment by sm-iitk** [comment]
As per the user, he did not put any specific preference, and used the defaults. Can you give me a hint which records are important how to get those, so that i can paste those records here. There are still few users for whom i have not fixed this issue yet.

**Comment by alecpl** [comment]
It might be anything, but I was thinking about the record from `users` table. Also, an information on plugins used should be provided in case of any issues.

**Comment by sm-iitk** [comment]
As per the user, he did not put any specific preference, and used the defaults. Can you give me a hint which records are important how to get those, so that i can paste those records here. There are still few users for whom i have not fixed this issue yet.

**Comment by sm-iitk** [comment]
This error is affecting only 1% user randomly.
one affected users table content (personal data removed)
https://drive.google.com/file/d/0B7PuQKEXBr6yVGZ5VWtXUUJxekk/view

I use lots of plugin, which works of all users. I shall do the plugin error check with the users.

**Comment by alecpl** [comment]
I really propose to disable all plugins and see if it can be still reproduced. Also, ask for the web browser used by users with this issue. Please, collect more info from users, does it happen quickly after login or after long usage of the same session? Does it happen on draft save or send action? Automatically or on user action?

The column where we store session data is MEDIUMTEXT, I don't think anyone exceeds limit of 16MB and probably it would destroy the whole session, but that's something to check.

**Comment by dbucher** [comment]
Hello I have the same problem since upgrading to version 1.2.5.

[21-Jun-2017 17:54:43 +0200]: <4276bc94> PHP Error: Invalid compose ID in /var/www/outils/roundcubemail-1.2.5/program/steps/mail/sendmail.inc on line 39 (POST /outils/webmail/?_task=mail&_unlock=loading1498060460543&_lang=fr_FR&_framed=1&_action=send)

How is it possible to help debbuging ? It happens almost most of the time.

Somewhere I saw it could be due to the number of message in Inbox ? I have 125840 emails and 11850 unread.

**Comment by dbucher** [comment]
P. S. ip_check = false in config/defaults.inc.php, and we are not logged out.

To be able to send the mail I need to create a new email in Inbox again. I had the problem with reply AND with new (blank) emails

**Comment by dbucher** [comment]
P. P. S. We only have 4 plugins :     'archive',    'zipdownload',    'managesieve',    'mobile',

**Comment by derproblembaer** [comment]
I already commented in issue #5390 

Our users still experiencing the same problem. It happens with all 1.2 versions, with 1.1 it's perfectly fine.

I tried it without plugins, Installed it on a whole new server with seperate database and created a new config from the ground up without success.

Tested Setups:
Debian 8.8.0, 9.0.0
Apache 2.4.10,  2.4.25
PHP 5.4.45 + 7.0

**Comment by dapphp** [comment]
I just wanted to weigh in here and report that we've been seeing this happen with various users as well.

In the last 2 weeks, I count it happening 709 times out of 77,328 successful messages.  Some of these look like people pressing send twice really fast (but not all).  I was able to slightly reproduce the error by pressing send rapidly (it sent successfully and I generated two consecutive compose ID errors, but since the first send was successful, I only briefly saw the error window before being redirected to the inbox.  So I think some of the 709 are this (due to the close proximity of a successful send and an error, where others are naturally occurring).  Users also report not getting logged out.  They can still do other things, only an issue composing.

We're running a 2 server load-balanced set up (both share a DB server) and it happens about equally between the two different servers.

For well over a couple months we've been on 1.2.5 and have been seeing it there.  Going to have some users beta the 1.3.0 release and see if it continues and report back.
